### ISSUE TITLE
API to resolve language version of a package

### ISSUE THREAD

**Issue by simolus3** [dart-lang/build#3492]
I have a builder that generates Dart files without having Dart sources as its input. These generated sources should respect the language version of the package for which they're generated (for instance, I can generate record types if a user depends on `>=3.0.0-0` and fall back to classes otherwise).
When there is a primary Dart input file, we can use the analyzer API to resolve its language version. When code is generated for other assets though, there is no easy way to determine a package's language version. (Using build APIs to read a pubspec works, but requires builder authors to replicate a lot of logic already handled by first-party Dart tools. `Isolate.packageConfig` also works, but might break hermetic builds)

I would like to have a method on `Resolver` that returns the language version of a package, e.g. 

```dart
  /// Returns the default Dart language version for sources in the given
  /// [package] that don't have an explicit [LanguageVersionToken] override.
  ///
  /// Throws a [PackageNotFoundException] if the package is not part of the
  /// build.
  Future<Version> languageVersionForPackage(String package);
```

I'm happy to help implement this of course (and it seems like we have all the necessary information in `build_resolvers`), but it would be a breaking the `Resolver` interface and it doesn't seem like this is worth a build 3? Previously the `build_resolvers` package would depend on minor versions of the `build` package, but that seems to have changed. So if this API proposal is accepted, I could use some advice on how to do it with minimum breakage (maybe put it on `BuildStep` instead?).

**Comment by jakemac53** [comment]
We have had previous asks for exposing the entire package config more generally, which would allow this plus potentially some other things. So I would probably take that approach, and hang it off of the `BuildStep`.

As far as invalidation goes I believe we already invalidate the whole build if the package config changes at all so it should be handled. But we could also just treat the package config as a source file (although it would likely need some special handling since I think generally we ignore hidden dirs).

**Comment by jakemac53** [comment]
Fwiw, it would also be generally safe to just use `Isolate.packagConfig` today, since we will invalidate the build if a language version changes etc.

**Comment by simolus3** [comment]
Yeah, I've also used `Isolate.packageConfig` as a workaround. I just got worried because the config can be configured for the resolvers, so I was afraid there might be some build setup where that would yield incorrect data.

I can expose the `PackageConfig` from `package:package_config` in a build step as is. That would probably be the easiest solution, but it also feels a bit weird because the file URIs in `Package.packageUriRoot` make it very easy for builder authors to break builds. Do you recall the other requests for this (if they're public)? It seems like the only helpful information we're getting from a full package config is an enumeration of all packages, but maybe I'm missing something.

**Comment by simolus3** [comment]
Or we can just hide the package URIs from builders by replacing them with `asset:` URIs, that would make it fairly safe to expose the full package config I think.

**Comment by jakemac53** [comment]
Hmm good point, it might encourage people to do bad things. Technically people can already use the `Isolate.resolvePackageUri` (or the `Isolate.packageConfig`) apis to get this, but if we expose it directly that is a little more sketchy.
